PPM-249 - independent unit to deal with hostapd configuration

RanRegev opened the pull request:
Encapsulates `hostapd` configuration separately.
The separation enables:
* reuse in more than one place: e.g. nl80211, dwpal, mock
* testing
* independent development
For more detail:
This PR: https://jira.prplfoundation.org/browse/PPM-249
General Overview: common/beerocks/hostapd/README.md

RanRegev committed:
prplmesh-hostapd: info: added README.md file

A file explaining the content of this directory
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added configuration class

added hostapd Configuration class to
be used by applications that wants to
edit, add, delete, enable, disable, etc
parametrs in hostapd configuration file
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added a unit test file

unit tests to test hostapd
configuration functionality
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added cmake for prplmesh hostapd

created CMakeLists.txt to build prplmest_hostapd library
and its unit tests
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

arnout commented on common/beerocks/hostapd/INFO.md:
```diff
@@ -0,0 +1,36 @@
+
```
If you call the file README.md instead of INFO.md, it will be displayed in the web interface automagically.

RanRegev replied:
done.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,23 @@
+
+namespace prplmesh {
+namespace hostapd {
+namespace config {
```
IMO adding a namespace per class is kind of useless... But we still need to finalize the namespace discussion.

RanRegev replied:
I wasn't sure it will end up with only one class, I thought that there might be more things in the namespace, and therefore there is a logic in adding this namespace.
e.g. assume that we have a free function named `reload()` in namespace `prplmesh::hostapd` - one can be confused if it is `config::reload()` or `process::reload()`.
Anyhow, as you mentioned, a farther discussion is needed.
I am leaving the namespace as is at the moment.

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,56 @@
+     * @brief construct a Configurationn object
+     * @param file_name the name of the configuration file.
+     */
+    Configuration(const std::string file_name);
```
```suggestion
    Configuration(const std::string& file_name);
```

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,56 @@
+     * }
+     *
+     */
+    void load();
```
```suggestion
    bool load();
```

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,23 @@
+namespace hostapd {
+namespace config {
+
+Configuration::Configuration(const std::string file_name) : m_configuration_file(file_name) {}
```
This becomes easier to test if instead of a filename, you give if an `istream&`. That way, you can pass in a stringstream in the test.

RanRegev replied:
indeed, and it is possible, and might even be a good idea.
However, from my experience, and since the file format is so delicate, and since the test flow is: load, change, **store**, **compare** - 
- we'll need an an `ostream` for storing (current file name serves both I/O).
- having the ability to load from a string may decrease the chances that one would compare the input to the output.

arnout commented on common/beerocks/hostapd/CMakeLists.txt:
```diff
@@ -0,0 +1,135 @@
+project(prplmseh_hostapd VERSION ${prplmesh_VERSION})
```
(nitpick) I prefer if the module name is the same as the directory name.

RanRegev replied:
@fvbogaert  - fixed once I searched for the test application and couldn't find it by name :smile:.
@arnout - I thought about it and came to the conclusion that `hostapd` would be _extremely_ confusing on one hand, and naming the directory `prplmesh_hostapd` would be odd. I couldn't end up with something good, so I am open for better naming.

arnout commented on common/beerocks/hostapd/CMakeLists.txt:
```diff
@@ -0,0 +1,135 @@
+set(MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
+
+# Support more permissive c++ language
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11")
```
I don't think you're going to use GNU extensions, so don't do this.

RanRegev replied:
Removed

arnout commented on common/beerocks/hostapd/CMakeLists.txt:
```diff
@@ -0,0 +1,135 @@
+    find_path(PRPLMESH_HOSTAPD_INCLUDE_DIR NAMES "src/common/wpa_ctrl.h" PATHS ${PRPLMESH_HOSTAPD_SEARCH_PATHS} NO_CMAKE_FIND_ROOT_PATH)
+    set(PRPLMESH_HOSTAPD_DIR "${PRPLMESH_HOSTAPD_INCLUDE_DIR}")
+
+endif()
```
None of this is needed. If it turns out to be needed, introduce it when you need it.

RanRegev replied:
Removed
Will be introduced when needed.

arnout replied:
It's still there.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -16,8 +19,122 @@ Configuration::Configuration(const std::string file_name) : m_configuration_file
+    m_ok = parsing_vaps && !ifs.bad();
+}
+
+std::ostream& operator<<(std::ostream& o, const Configuration& conf) {
```
If you make this
```suggestion
template<class T> T& operator<<(T& o, const Configuration& conf) {
```
it will also work as `LOG(DEBUG) << conf;`

RanRegev replied:
[mimic :scream:](https://github.com/prplfoundation/prplMesh/blob/master/framework/tlvf/src/include/tlvf/tlvftypes.h#L71)

arnout replied:
Hm, I wrote that code, so I guess it didn't work with the template since I'm sure I considered it at the time...

RanRegev committed:
prplmesh-hostapd: added set_create_vap_value()

introducing: set_create_vap_value
functionality and unit test
also: few minor general issues:
* a variable for the test file
* interface to get the last message (not implemented yet)
* change the variable name from error to message and fix
accordingly in the code
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added set_create_vap_value() overload with int

introducing: set_create_vap_value, int overload
implemented the function using the string version
added functionality and tests
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added store()

introducing: store
stores internal configuration into the same
file it was loaded from, effectively changing the
configuration.
added functionality and tests
also:
* chnaged prerequisite in tests from
EXPECT to ASSERT - as we can't continue without it
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added get_vap_value()

introducing: get_vap_value
added the ability to get a value
for a specific vap with a given key.
added functionality and tests
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added disable_all_ap_vaps()

added functionality to disable all vaps that
are access-points (as opposed to STA for example).
the code mimics the existing functionality taken from
ap_wlan_hal_dwpal.cpp.
added funcvtionlity and unit test
note:
the example unit-test configuration file did not
contain any vap with the mode= line, therefore it
was added few of this.
also:
it was discovered that it is OK to look for non-exiting key
within vap's configuration and the code was changed accordingly
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added private help functions

added private helper functions for functions
changed all usages accordingly
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added comment_vap()

comment out (disable) the given vap
by adding a comment (#) to it and to
all keys related to this vap
added functionality and unit test
also:
clang format chages
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added comment ability

load, search and store comented lines in the
file. this ability is needed for disabling
vaps in the "up stream" version of hostapd.
the intel version disables vaps by adding the line
"start_disabled=1"
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added uncomment_vap()

uncomment (enable) the given vap
by removing all comment signes (###) from it
and from all keys related to this vap
added functionality and unit test
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added for_all_ap_vaps()

introducing: a function that goes all over all
vaps which their mode is "ap" and apply the
input function to them.
may be used for example to disable all
vaps.
added functionality and tested by:
- adding a parameter "start_disabled=1" to all.
- commenting out
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added for_all_ap_vaps() with user iterator

There is a need to go on both the ap-vaps configured in hostapd
and in paralell go over the given configuration (non-hostapd structure)
The added functionality gives hostapd configuration class the
responsibility of iterating over the vaps _and_ increment the
user-iterator. the result is that the user's function is called
with the next vap _and_ the next element in its given list.
it is the user responsibility however to make sure that its container
has at least as the number of vaps. otherwise for_all_ap_vaps() would
increment the iterator passed the end of the user's container
tested by implementing the non iterator version of for_all_ap_vaps()
in term of the iterator version, with a dummy iterator
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added test enable_all_vaps()

a test that goes all over all
vaps and enables them
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added test iterate-both-containers

testing the ability to iterate over the user
container and the hostapd's ap-vaps in paralel
giving the user the ability to set parametrs
based on its own container and the hostapd container
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: few cleanups after rebase

rebase --autosquash -i origin/master
introduced issues that I couldn't figure out
how to fix with rebasing again.
gave up and added this fix commit.
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added set_create_xx_head_value

added a functionality to set or create key/value
in the header part: either string or int
added functionality and tests
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added get_head_value()

added a function to get a value from the head section
added functionality and unit test
also in this commit:
the variable named it_str is now called line_iter
(it is better describs its meaning)
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

mariomaz commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+
+    while (getline(ifs, line)) {
+
+        // skip empty lines
```
Should you remove leading and trailing white spaces before checking for empty line?

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+
+namespace {
+
+const std::string configuration_path("/tmp/");
```
Don't use `std::string` for string constant expressions - use `const char*` instead as it's more efficient.

RanRegev replied:
1. references?
2. why do we need efficiency in this area?
3. I kindly refuse to deal with efficiency/performance issues without measurements.
4. I prefer code safety and correctness unless other forces are stronger (binary size / performance / etc)

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+    //// start prerequsite ////
+
+    // save the content of the string (start clean)
+    //   // clean_start();
```
Remove commented code

mariomaz replied:
Tests should not depend on the order of execution. GoogleTest has a command line argument to [shuffle](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#shuffling-the-tests) the tests (it runs the tests in a random order). GoogleTest has also a command line argument to filter the tests to be run (so dependencies might not have been satisfied).

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+    //// end prerequsite ////
+
+    // add a value to vap
+    conf.set_create_vap_value("wlan0.3", "was_i_stroed", "yes_you_were");
```
```suggestion
    conf.set_create_vap_value("wlan0.3", "was_i_stored", "yes_you_were");
```

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+void clean_start()
+{
+    // save the content of the string (start clean)
+    std::ofstream tmp(configuration_path + configuration_file_name);
```
If you use the same file name for all the tests, then you cannot run them in multiple simultaneous threads. It's better to choose a different file name for each test.
To avoid file name conflicts in case more than one instance of the test application was executed in the same system at the same time, I used to use UUIDs for file names.
You should read the TMPDIR environment variable to get the name of the temporary directory instead of assuming it's "/tmp"

arnout replied:
To be completely accurate, you should use mkstemp() - that guarantees a unique file.

However, even better would be if the file itself could be mocked away. I tried to convince @RanRegev of this but no success. So, I guess it should be merged as is and @mariomaz can have fun with creating something that uses mocks.

Note that *all* of Mario's other comments to this unit test would be resolved if the file were mocked away...

RanRegev replied:
I was convinced, and will eventually be aligned with this. I have two major issues that hold me from doing it now:
1. Time frame. I need some ramp up to learn and work with these MOCKS.
2. From the little I saw - I would need to have an interface for `File` **just for** the sake of the tests. From my point of view, if I write a `File` interface, it should be an interface that satisfies _many_ kinds of files, including, among others, the _mock_ files. This is entirely different mission than just creating a `File` interface that is good for my current needs and tests. In other words, thinking only on current needs and mocking will eventually resulted with something that is usable in these two use cases _only_. This is not an interface as I see it. Interface for me is something that I can use in many places, usually written for one or two existing use cases - but is truly opened for farther extending.

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+{
+    //// start prerequsite ////
+
+    clean_start();
```
If there's code that must be executed at the beginning or at the end of each and every test, define a test fixture for that.
If the test creates a file (even in the temporary folder), the test should remove it when done.

mariomaz commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -59,18 +59,66 @@ bool Configuration::load()
+    }
+
+    // when the new value is provided add the key back with that new value
+    if (value.length() != 0) {
```
```suggestion
    if (!value.empty()) {
```
Swap if-else clauses and do not use a negated condition

mariomaz commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -59,18 +59,66 @@ bool Configuration::load()
+    // we push it to the end of the array
+
+    // delete the key-value if found
+    if (it_str != vaps_values.end()) {
```
If possible, do not use negated conditions (code is clearer to read)

mariomaz commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -48,6 +48,24 @@ class Configuration {
+     * true - the vap exists, values were set.
+     * false - the given vap was not found
+     */
+    bool set_create_vap_value(const std::string &vap, const std::string &key,
```
Method should include one verb only, either "set" or "create", but not both

mariomaz commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -48,6 +48,24 @@ class Configuration {
+     * message (similar to errno) - for debug usage
+     * @return s string describing the last message
+     */
+    const std::string &get_last_message() const;
```
```suggestion
    const std::string &get_last_error_message() const;
```

mariomaz commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -79,7 +97,7 @@ class Configuration {
     std::map<std::string, std::vector<std::string>> m_hostapd_config_vaps;
 
-    std::string m_error_description = "no error";
+    std::string m_last_message = "all good";
```
```suggestion
    std::string m_last_message = "";
```
Use empty message for no error

mariomaz commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -114,6 +115,14 @@ bool Configuration::set_create_vap_value(const std::string &vap, const std::stri
 }
 
+bool Configuration::set_create_vap_value(const std::string &vap, const std::string &key,
+                                         const int value)
```
```suggestion
                                         int value)
```

arnout replied:
We don't normally use `int` for anything... We should probably use the specific types that are actually used (mostly uint8_t or uint32_t, probably), or use `long long` and `unsigned long long` as generic types that can take any int without loosing bits.

But since this is only removing a `std::to_string`, and the caller anyway has to be very aware of what exactly hostapd.conf expects (maybe it's not an int but there's still some suffix), I don't think it's worth having this function to begin with.

But I see that this was already like this in the original, so OK.

mariomaz commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -68,14 +68,24 @@ class Configuration {
+
+    /**
+     * @brief get the value of the given key for the given vap
+     * @param 
```
Is this a valid doxygen format for `@param`s?

arnout replied:
Nope.

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -535,8 +535,31 @@ TEST(configuration_test, get_vap_values)
+}
+
+
+TEST(configuration_test, diable_all_ap)
```
```suggestion
TEST(configuration_test, disable_all_ap_vaps)
```

mariomaz commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -576,4 +576,45 @@ TEST(configuration_test, disable_vap)
+    conf.uncomment_vap("wlan0.1");
+    EXPECT_TRUE(conf) << conf;
+
+    conf.uncomment_vap("wlan0.2");
```
These tests are not actually testing, they are just running the code

mariomaz commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -150,6 +150,23 @@ bool Configuration::set_create_head_value(const std::string &key, const int valu
+    if (line_iter == m_hostapd_config_head.end()) {
+        m_last_message = std::string(__FUNCTION__) +
+                         " couldn't find requested key in head: " + key;
+        return "";
```
Shouldn't you set `m_ok = false` before returning?
Also, if function succeeds, `m_last_message` should be cleared.
If `m_last_message` and `m_ok` are updated in parallel, you might infer `m_ok` from `m_last_message`

mariomaz reviewed (changes requested):
Unit tests should be improved.

arnout commented on common/beerocks/hostapd/README.md:
```diff
@@ -0,0 +1,36 @@
+Manages hostapd process
+* start
+* stop
+* restart
```
We don't start or stop hostapd.

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * @brief construct a Configurationn object
+     * @param file_name the name of the configuration file.
+     */
```
```suggestion
    /**
     * @brief Construct a Configuration object.
     *
     * @param file_name the name of the configuration file.
     */
```

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    ~Configuration()                     = default;
+    Configuration(const Configuration &) = default;
+    Configuration(Configuration &&)      = default;
+
```
I like an explicit
```suggestion
    Configuration()              = delete;

```

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+     * }
+     *
+     */
+    bool load();
```
So, no RAII then?

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    bool m_ok = false;
+
+    // each string is a line in the original configuration file
+    // that belongs to the "head" part. read the explnation at
```
```suggestion
    // that belongs to the "head" part. read the explanation at
```

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    std::string m_configuration_file;
+    bool m_ok = false;
+
+    // each string is a line in the original configuration file
```
Even for private members, it makes sense to use doxygen formatting:
- doxygen by default doesn't generate documentation for private members;
- it's anyway useful to have such documentation for developing the class itself;
- it simplifies your life when later you decide to make the member public/protected.

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    friend std::ostream &operator<<(std::ostream &, const Configuration &);
+};
+
+// for logs
```
Well, not actually for logs, because elpp logs are not ostream...

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+
+Configuration::Configuration(const std::string& file_name) : m_configuration_file(file_name) {}
+
+Configuration::operator bool() const { return m_ok; }
```
(nitpick) I prefer to define functions like this inline in the class.

It does make the header file larger, but myself I hardly ever use the header file directly for documentation: I use my IDE or doxygen output.

And the advantage is that you can kill any comment of the type "this will make it less efficient", because the compiler will be able to optimise it all away. If you hide the implementation in the cpp file, it's a lot more difficult to optimise it (other than with LTO but that's a different can of worms).

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+    std::ifstream ifs(m_configuration_file);
+    std::string line;
+
+    bool parsing_vaps = false;
```
(nitpick) this is redundant, `!cur_vap.empty()` works as well.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+        // skip empty lines
+        if (std::all_of(line.begin(), line.end(), isspace)) {
+            continue;
+        }
```
Maybe it was already like that in the original code, but I don't like this very much. We should reproduce the original hostapd.conf file as much as possible, including empty lines, comment lines, etc.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+        if (!parsing_vaps) {
+            m_hostapd_config_head.push_back(line);
+        } else {
+            m_hostapd_config_vaps[cur_vap].push_back(line);
```
Since there's no `continue` above, the `bss=wlan0.1` line will be included as the first line in `m_hostapd_config_vaps[cur_vap]`. However, the documentation in the header file doesn't mention this.

arnout commented on common/beerocks/hostapd/CMakeLists.txt:
```diff
@@ -0,0 +1,85 @@
+
+endif()
+
+set(PRPLMESH_HOSTAPD_TYPE ${PRPLMESH_HOSTAPD_TYPE} CACHE STRING "Which PRPLMESH_HOSTAPD backend to use")
```
This is also not used yet.

When you use it, please make the values lower case so you don't have that silly conversion to lowercase just to get a usable name.

arnout commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+ * See LICENSE file for more details.
+ */
+
+#include "configuration.h"
```
```suggestion
#include <configuration.h>
```

arnout commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+
+#include <gtest/gtest.h>
+
+namespace {
```
For my information: why put this in an anonymous namespace, rather than the same prplmesh::hostapd::config namespace as the implementation?

arnout commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+
+const std::string configuration_path("/tmp/");
+const std::string configuration_file_name("hostapd.conf");
+const std::string configuration_content(
+    "driver=nl80211\n"
+    "logger_syslog=127\n"
+    "logger_syslog_level=2\n"
+    "logger_stdout=127\n"
+    "logger_stdout_level=2\n"
+    "sDisableMasterVap=1\n"
```
Maybe it's just me, but I prefer unit tests to use simple values that show exactly the corner cases we want to handle. So I'd expect a test with empty content, with missing `bss=` line, with empty lines, with whitespace before the variable name, etc. etc. But I would make each one very small and simple.

An actual, realistic configuration file *will* be tested by the integration test, so I don't think it's needed to do that in the unit test. Unit test, IMHO, is for corner cases.

arnout commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+
+    // load the dummy configuration file
+    conf.load();
+    ASSERT_TRUE(conf) << conf;
```
I would use a different Configuration object here (especially of course since I think load() should be part of the constructor). And also, I'm missing tests that validate that the configuration file was written correctly. It would help if you have a mock for the file, then you could just add expectations on the writeline() calls :-)

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -68,6 +68,36 @@ bool Configuration::load()
+    for (auto &vap : m_hostapd_config_vaps) {
+
+        // add empty line for readability
+        out_file << "\n";
```
As written elsewhere, I'd prefer to keep the exact same formatting as the original file.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -68,6 +68,36 @@ bool Configuration::load()
+    }
+
+    // store the vaps
+    for (auto &vap : m_hostapd_config_vaps) {
```
So, this is why I don't like maps... Both map and unordered map will not iterate over the vaps in the original order. Instead, they'll take some order as defined by the internal sorting of the container. And unfortunately, this reordering *may* have some unwanted side effects - hostapd.conf is weird.

I'd really prefer a vector for m_hostapd_config_vaps.

arnout commented on common/beerocks/CMakeLists.txt:
```diff
@@ -1,10 +1,11 @@
-add_subdirectory(tlvf)
 add_subdirectory(btl)
+add_subdirectory(bwl)
+add_subdirectory(hostapd)
```
Commit message says "created CMakeLists.txt" but the commit *actually* adds the CMakeLists.txt created in the previous commit to the parent, and sorts the add_subdirectory calls alphabetically. The latter should be mentioned in the commit message as well.

But there's really no need to split off the creation of CMakeLists.txt or adding it to the parent in a separate commit.

Similarly, when a unit test is added, it's better to add it to CI in the same commit.

By the way, you may want to re-read https://github.com/prplfoundation/prplMesh/blob/master/CONTRIBUTING.md#commits. About the subject line:
* Write it in the imperative: "add support for X".
* Avoid verbs that don't say anything: "fix", "improve", "update", ...

And about the body:
* the (core) change itself - in the imperative (i.e. "add foo()", not ("added foo()");

And it's not written explicilty there, but the examples show that the JIRA URL should be a separate paragraph.

All of this is not very important, i.e. not worth spending your time on fixing all the commit messages are commits. Just something to take into account for the future.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -59,18 +59,66 @@ bool Configuration::load()
+    auto existing_vap = m_hostapd_config_vaps.find(vap);
+
+    if (existing_vap == m_hostapd_config_vaps.end()) {
+        m_last_message = std::string(__FUNCTION__) + " couldn't find requested vap: " + vap;
```
Why do we store error messages for consumption by the caller, instead of simply logging it directly?

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -59,18 +59,66 @@ bool Configuration::load()
+    } else {
+        m_last_message =
+            std::string(__FUNCTION__) + " the key '" + key + "' for vap " + vap + " was not found";
+    }
```
```suggestion
    } else {
        LOG(TRACE) << "The key '" << key << "' for vap " << vap << " was not found";
    }
```

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -59,18 +59,66 @@ bool Configuration::load()
+            std::string(__FUNCTION__) + " the key '" + key + "' for vap " + vap + " was deleted";
+    }
+
+    m_ok = true;
```
That's not good... There's no check of m_ok at the beginning of this function, so we can magically repair a broken situation by setting a value...

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -48,6 +48,24 @@ class Configuration {
 
+    /**
+     * @brief set key/value for the given vap
+     * @details set the key/vale for the given vap, either replace or create.
```
Or remove... Which is something I don't really like. I'd prefer a separate remove function. But (nitpick).

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -151,6 +151,37 @@ bool Configuration::set_create_vap_value(const std::string &vap, const std::stri
+    if (existing_vap == m_hostapd_config_vaps.end()) {
+        m_last_message = std::string(__FUNCTION__) + " couldn't find requested vap: " + vap +
+                         " (requested key: " + key + ")";
+        m_ok = false;
```
Hm, I don't really like this... A missing VAP doesn't mean that the configuration suddenly becomes invalid. IMHO it would be more logical to return bool and have an output argument for the value. The caller will anyway need a condition to check if the key was present, so it's not possible to chain this function into another operation.

(As usual, it would be a lot easier if we could throw an exception...)

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -151,6 +151,37 @@ bool Configuration::set_create_vap_value(const std::string &vap, const std::stri
+                         " couldn't find requested key for vap: " + vap + "; requested key: " + key;
+        m_ok = false;
+        return "";
+    }
```
(nitpick) can be refactored with set_

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -182,6 +185,17 @@ std::string Configuration::get_vap_value(const std::string &vap, const std::stri
+                    set_create_vap_value(vap.first, "start_disabled", "1");
+                }
+            });
+}
```
This function is never going to be used. There is no bwl function to disable all VAPs, and there's no need for it.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -26,6 +26,11 @@ bool Configuration::load()
+    // for cases when load is called more than once, we
+    // first clear internal data
+    m_hostapd_config_head.clear();
+    m_hostapd_config_vaps.clear();
```
RAII -> only load once -> this is not needed.

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -97,10 +97,33 @@ class Configuration {
+     * @return tuple<bool,std::vector<std::string>*> (found/not found, a 
+     * pointer to the vap's array)
+     */
+    std::tuple<bool, std::vector<std::string> *> get_vap(const std::string &calling_function,
```
There's no need to return a tuple since the pointer takes the role of the boolean.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -154,35 +158,30 @@ bool Configuration::set_create_vap_value(const std::string &vap, const std::stri
+    if (!std::get<0>(find_vap)) {
         return "";
     }
+    const auto& existing_vap = std::get<1>(find_vap);
```
It's a pointer so const-ref is kind of silly.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -198,6 +197,43 @@ void Configuration::disable_all_ap_vaps() {
+    // - make sure it is either on the first position
+    // or it has a comment sign just before it
+    auto found_pos = line.rfind(key);
+    bool ret = found_pos != std::string::npos && (found_pos == 0 || line.at(found_pos - 1) == '#');
```
I think it would be better to remove all the comments when parsing, and to add a flag to the vap structure (which should really be a structure/class rather than a plain vector<string>) to indicate if it's disabled.

However, all this is not needed. Upstream hostapd supports start_disabled as well. So we're not going to comment anything out.

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -184,6 +191,16 @@ class Configuration {
+template <class func> void Configuration::for_all_ap_vaps(func f)
+{
+    for_each(m_hostapd_config_vaps.begin(), m_hostapd_config_vaps.end(),
+             [this, &f](const std::pair<std::string, std::vector<std::string>> &vap) mutable {
```
For my info: what is the meaning of the mutable keyword here?

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -123,6 +124,16 @@ class Configuration {
+     * in paralel: the ap-vaps and the user container.
+     */
+    template <class func, class iter> void for_all_ap_vaps(func, iter current, const iter end);
+
```
What we *actually* need is the possibility to iterate over all vaps in parallel with the function, and then perform another function on the remaining vaps.

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -192,11 +203,25 @@ class Configuration {
+    auto f_with_iter = [&f](const std::string &vap, int iter) { f(vap); };
+
+    int dummy(0);
+    for_all_ap_vaps(f_with_iter, dummy, 10);
```
Err, so max 10 vaps supported?

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -192,11 +203,25 @@ class Configuration {
+}
+
+template <class func, class iter>
+void Configuration::for_all_ap_vaps(func f, iter current_iter, const iter end)
```
I would have turned it around, something like this:
```suggestion
template <class func, class iter>
void Configuration::for_all_ap_vaps(func f, iter current_iter, const iter end)
{
    for_all_ap_vaps([](const std::vector<std::string> &vap) {
        if (current_iter != end) {
            f(vap, current_iter);
            current_iter++;
        } else {
            /* Call the second function here for the remaining vaps */
        }
        });
```

arnout commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -192,11 +203,25 @@ class Configuration {
                  if (get_vap_value(vap.first, "mode") == "ap") {
-                     f(vap.first);
+                     if (end == current_iter) {
+                         f(vap.first, end);
```
Ah, actually this way indeed you can handle the remaining ones... But it's clunky because you need the condition inside, a separate function would be nicer.

arnout commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -209,6 +209,10 @@ TEST(configuration_test, store)
+    EXPECT_FALSE(conf) << conf;
+
     // construct a configuration
     ASSERT_FALSE(conf) << conf;
```
expect then assert of the same is weird.

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -112,6 +112,44 @@ bool Configuration::store()
     return *this;
 }
 
+bool Configuration::set_create_head_value(const std::string &key, const std::string &value)
```
Again, we don't need this. We only modify VAPs.

arnout reviewed (approved):
As usual, I have a lot of comments on this PR.

The commit splitting is a bit of a mess. But I guess I rushed you a bit :-) So OK.

One problem is that it adds a lot of unneeded features. Please don't do that. We don't need commenting out, we don't need changing the header part, we don't need disable of all vaps. So almost half of the commits are not even needed.

That said, nothing is critically wrong. The only critical thing is the ordering of the VAPs that isn't maintained. However, it will work in practice (at least for the time being) since openwrt generates the config file with sorted bss= sections. Well, actually, you *can* do funny things in uci wireless so that it generates it in a different order. Anyway, I think it can be merged as is (maybe with some cosmetic changes like fixing the errors in doxygen) but a bug has to be created for the ordering of VAPs.

RanRegev committed:
prplmesh-hostapd: added vap_indication to load()

originally bss= was the only sign for
vap configuration inside the hostapd file.
this was based on intel's hostapd version.
the hostapd upstream indicates vap configuration
with interface= sign.
load() now seeks for the vap sign as set
by the user.
also in this commit:
- removed namespace config from prplmesh::hostapd.
left with only the two and class Configuration.
- clang format fixes
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added ap predicate to for_all_ap_vaps

there is a need to let the usr the ability to decide
if a given vap is an ap vap or not.
e.g.
for intel the predicate is: mode=ap
for up stream's hostapd: all vaps are ap vaps

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
prplmesh-hostapd: added ap predicate to for_all_ap_vaps

there is a need to let the usr the ability to decide
if a given vap is an ap vap or not.
e.g.
for intel the predicate is: mode=ap
for ap stream hostapd: all vaps are ap vaps
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
nl80211: hostapd configuration: cmake

added prplmesh_hostapd dependency to bwl and btl
in cmake

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
nl80211: hostapd configuration update_vap_credentials()

added hostapd configuration to
update_vap_credentials()
the implementation was copied and then adopted
from dwpal implementation
(common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp)
tested on Turris Omnia

Signed-off-by: Ran Regev <ran.regev@devalore.com>

arnout commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,390 @@
+between /// BEGIN hostapd.conf /// and /// END hostapd.conf /// is the 
+format of the file.
+note: the string "bss=" may be replaced by the
+user in the call to load() with another vap-indicator (e.g. "interface=")
```
This is wrong. The actual format of the file is:
```
... Configuration of first VAP
bss=...
... Configuration of second VAP
bss=...
... Configuration of third VAP
```
`interface=` sets the name of the VAP. `bss=` also sets it, so it really is only needed for the first VAP.

As it happens, openwrt generates the configuration file as:
```
... Global options
interface=first_vap_name
... First VAP options
bss=second_vap_name
... Second VAP options
```

But that's just "coincidence", not really something we can rely on...

RanRegev committed:
prplmesh-hostapd: PR 1514 review fixups

this commnit fixes all relevant comments for
https://github.com/prplfoundation/prplMesh/pull/1550
https://jira.prplfoundation.org/browse/PPM-249

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
nl80211: hostapd configuration: close issues found in PR 1550

this commit closes issues found in
https://github.com/prplfoundation/prplMesh/pull/1550
https://jira.prplfoundation.org/browse/PPM-4
MAP-4.2.1:turris-omnia

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev commented:
Replaced by #1601
